### PR TITLE
[WIP]: Sanatize ids during getFeature

### PR DESCRIPTION
**Pull request by loicgasser**
With the new geocover "shop" update we have primary keys of different types.

For some models the id is some text and for some other we have integers...

This is to handle the case where we would request a table using the wrong type.

I would add that this is a temporary and I have a greater scheme to handle all these border line cases.

@gjn Please have a look.

### CODE REVIEW

**Comment by gjn** [comment]
Can't we integrate the data in such a way that both primary keys have the same type (string in this case?). The import should also assure that it's indeed unique accross multiple tables (as the service assumes as much).

File: chsdi/views/features.py [hunk 1]
```diff
@@ -318,21 +318,33 @@ def _get_features(params, extended=False):
             yield _get_feature_db(featureId, params, models)
 
 
+def is_valid_id_value(column, value):
+    column_type = column.type
+    python_type = column_type.python_type
```
Comment (gjn) [inline chsdi/views/features.py]: Where is this coming from? (From sqlAlchemy?)

**Comment by loicgasser** [comment]
Yes

**Comment by loicgasser** [comment]
We should though if I create a request from scratch I can end up on a 500 error.

**Comment by gjn** [comment]
Yes, but that's already the case for other, 'correct' tables, no?

I'd rather adapt the data-integration. Can you talk about it with @pfanguin

**Comment by loicgasser** [comment]
Looking into that.

**Comment by loicgasser** [comment]
Okay so we can't change the ids of the grid layers as they are used by the toposhop.

We then then tried to use a hack by casting bgdi_id for the 7+ views associated to
 
ch.swisstopo.geologie-geocover
ch.swisstopo.geologie-geologischer_atlas

performance becomes really really terrible if you don't provide the scale (which is the case when you click a feature in the search) when requesting a feature in the grid. (35 sec)

The backend basically goes through 7+ views trying to match `GC-V-28` against a bgdi_id casted to text on the fly. -> totally unacceptable.

What's your take on that?

FYI: Looking at Jenkins I can tell that we have a lot "bad" mapping or "bad" tests.

**Comment by ltclm** [comment]
Why do we use these views for the shop?

**Comment by loicgasser** [comment]
I guess because we want to be able to order a digital map using some kind of custom pk25 grid.

I don't have time to find a better solution for tomorrow so I propose that we publish these 2 layers without the feature search (easily handled via layers configuration) warn the webshop people and find a proper solution for next deploy.

Regular Tooltip is still avaiable which is I believe the most important thing.

**Comment by ltclm** [comment]
Yes, that should be enough for tomorrow, thanks loic.
if we use these views for the shop ***, we are performing queries on more than a million features to get the map sheet id's of 222 objects - complete nonsense in my eyes. 
We have a geocover map sheet layer and model in chsdi which can be used for the shop link and a geologischer atlas map sheet layer too, we should use these meta layers instead of the data layers. 

https://map.geo.admin.ch/?X=128814.02&Y=650797.82&zoom=0&lang=en&topic=ech&bgLayer=voidLayer&layers=ch.swisstopo.geologie-geocover.metadata,ch.swisstopo.gas25-papierkarte.metadata&layers_opacity=0.75,1

**Comment by gjn** [comment]
Agree with @ltclm.
It's a data integration issue. Let's disable feature search and assure the data is well modelled and intgrated for our uses case.
Thanks for the analysis, @loicgasser.
Am 26.04.2016 9:08 nachm. schrieb Clausen Marcel <notifications@github.com>:Yes, that should be enough for tomorrow, thanks loic.
if we use these views for the shop ***, we are performing queries on more than a million features to get the map sheet id's of 222 objects - complete nonsense in my eyes. 
We have a geocover map sheet layer and model in chsdi which can be used for the shop link and a geologischer atlas map sheet layer too, we should use these meta layers instead of the data layers. 

https://map.geo.admin.ch/?X=128814.02&Y=650797.82&zoom=0&lang=en&topic=ech&bgLayer=voidLayer&layers=ch.swisstopo.geologie-geocover.metadata,ch.swisstopo.gas25-papierkarte.metadata&layers_opacity=0.75,1

**Comment by loicgasser** [comment]
@gjn Why don't we use ch.swisstopo.geologie-geocover.metadata intead of ch.swisstopo.geologie-geocover?
For the geocover stuff, we already have everything we need with this layer. It's because it's a digital product? I do not understand the logic behind it.

**Comment by gjn** [comment]
I have no idea. @p1d1d1 ?

**Comment by p1d1d1** [comment]
so has been defined by data owner.

**Comment by ltclm** [comment]
This layer does not contain the map sheet geometry and its primary key cannot be used as toposhop id. can we discuss this with the data owner?

**Comment by p1d1d1** [comment]
Has been already discussed several times. They want the shop functionality on the pixel layer.
@ltclm the layer has 2 geometry layers for the tooltip... and this is scale dependent.
For zoom levels 0-5 (I guess) it uses the .metadata table (I guess), for > 5 the table with the vector features is used.

**Comment by ltclm** [comment]
@p1d1d1 the shop link should only appear on the metadata table zoomlevels?

Do we need feature search on all zoomlevels or only the metadata zoomlevel?
If we need it on all zoomevels, i propose to add a new attribute to the feature tables and views which is ``NONE``  and which contains the Shop Key in the Metadata Shop Views. The current primary key of the feature tables should be removed from the queryable attributes, it is an autogenerated serial number!

**Comment by ltclm** [comment]
As discussed with @p1d1d1 the feature search should be activated ONLY for the metadata zoomlevels ``GeocoverGridShop ``. @loicgasser is this doable?

**Comment by AFoletti** [comment]
After a brief discussion with @ltclm and @rfrancel , our proposition is to handle that at BOD level and make the layer an aggregate.
This will allow to separate model classes assigns between mapsheet geometries and features geometries, thus solving the problem with a FSDI standard solution and without any hack

**Comment by loicgasser** [comment]
Aggregate sounds good but we're always requesting the parent layer in geoadmin.
In CHSDI the concept of aggreagted layers does not really exist. 
Parentlayer id is always used and the only way to interogate only part of a layer is to provide a scale (which you don't want to provide when clicking a feature from the search)

Now we could extend this behaviour backend, model layers group which would give us more flexibility.

If only parent layer id is provided -> look though all the models (default behaviour)

If scale is provided (e.g. MapExtent and MapImageDisplay) -> query only models at the appropriate scale

If parentLayerId and Sublayerid(s) are provided -> query all sublayers only.

It's quite a developement but it might be necessary as we are handling more and more diverse models (grid, toposhop extents) and the actual solution is not ideal. 
Your thoughts?

**Comment by p1d1d1** [comment]
Please consider that what you are discussing here applies also to ch.swisstopo.geologie-geologischer_atlas

**Comment by loicgasser** [comment]
This can be closed. I'll open a ticket for the issue we talked about.

**Pull request closed without merging**